### ISSUE TITLE
Add code fix to use another literal for CS8313, CS8363 A default literal 'default' is not valid as a case constant / pattern.

### ISSUE THREAD

**Issue by Neme12** [dotnet/roslyn#30360]
Related to #25557 but instead of a code fix to use a default expression, there should also be a code fix to use another literal (such as 0 or null), as described in the error message.

**Comment by alrz** [comment]
Could we make this a fix that compute some constant to the actual value?
```
int x = default;
int x = 0;
 -> 0
const int c = 0;
int x = c;
int x = 0;
```

**Comment by Neme12** [comment]
@alrz That's a good idea for a refactoring, but please file that as a separate issue. This is specifically a code fix for the compiler error you get when you have a default literal (so there isn't even much to compute here) in the wrong place.

**Comment by Neme12** [comment]
On another note, I think `default` *is* the actual value. So I like your refactoring proposal to inline constants, but I don't think it should be available on a default literal. Instead, there would be a separate refactoring to convert between a default literal and another literal (e.g. 0 or null), in both directions.

**Comment by sharwell** [comment]
We had a design review today, and concluded that for replacing default literals with another value, it is preferable to use the literal form when available in the language. This code fix *may* also be offered as a refactoring for default literals (when supported by the language) and default expressions (always) for cases where a literal form is available. The refactoring does not need to be implemented as part of the initial pull request (it can be a follow-up issue).

The following special cases were covered as well. Each of these special cases may be part of the initial implementation, or left to a follow-up issue.

* If the type is an enum type `T` with `[Flags]`, and the enum type defines a member `None = 0`, the literal value `T.None` would be preferable to `default(T)`.
* If the type is `CancellationToken`, `CancellationToken.None` is preferable to `default(CancellationToken)`.
* If the type is `IntPtr` or `UIntPtr`, the value `IntPtr.Zero` or `UIntPtr.Zero` is preferable to `default(IntPtr)` or `default(UIntPtr)`.

**Comment by Neme12** [comment]
@sharwell Just so I understand, there will be one code fix that will convert to `default(T)` *unless* there's an available literal, in which case it will generate the literal instead, *unless* it is any of the 3 cases above, in which it will apply the appropriate special case fix?

**Comment by sharwell** [comment]
> *unless* it is any of the 3 cases above, in which it ~~will~~ **may** apply the appropriate special case fix?

I fixed the last clause. The behavior is acceptable either way for that.

Also note that we found `default(short)` to be preferable to `(short)0` (there are a few types where this occurs.

**Comment by Neme12** [comment]
@sharwell I don't understand why there are so many special cases dedicated to handling code that would end up being illegal. `IntrPtr.Zero`, `UIntPtr.Zero`, `CancellationToken.None` are not constant values and therefore not allowed in a switch case or a constant pattern. My original `default` -> `default(T)` handled non-constants too because it didn't take any additional work, and it didn't seem harmful. It doesn't seem useful either though.

**Comment by Neme12** [comment]
> Also note that we found default(short) to be preferable to (short)0

Why not simply `0`?

**Comment by sharwell** [comment]
> > Also note that we found default(short) to be preferable to (short)0
> 
> Why not simply `0`?

Sometimes `0` is not possible, e.g. overload resolution or generic type inference. `0` would be an acceptable substitute if you verify it doesn't change the resulting semantics.

**Comment by Neme12** [comment]
@sharwell This code fix is for an invalid default literal in a switch case or a pattern. In those cases, there is always a target type (of the default literal), so 0 is possible because it will be converted to a short.

**Comment by sharwell** [comment]
@Neme12 Yes, for this case (and I believe all cases involving a default *literal*) it would simply allow `0`. The same general statement applies to #25556, and there it won't necessarily allow the use of `0`.
